test/src/test/java/jenkins/model/JenkinsTest.java

```diff
@@ -471,4 +480,116 @@ public void onOnline(Computer c, TaskListener listener) throws IOException, Inte
+                defaultProtocols, not(hasItem("JNLP3-connect")));
+        
+        final Set<String> newProtocols = new HashSet<>(defaultProtocols);
+        newProtocols.add("JNLP3-connect");
+        j.jenkins.setAgentProtocols(newProtocols);
+        j.jenkins.save();
+        final Set<String> agentProtocolsBeforeReload = j.jenkins.getAgentProtocols();
+        assertThat("JNLP3-connect must be enabled before the roundtrip", 
+                j.jenkins.getAgentProtocols(), hasItem("JNLP3-connect"));
+        
+        j.jenkins.reload();
+        
+        final Set<String> reloadedProtocols = j.jenkins.getAgentProtocols();
+        assertFalse("The protocol list must have been really reloaded", agentProtocolsBeforeReload == reloadedProtocols);
+        assertThat("We should have additional enabled protocol", 
+                reloadedProtocols.size(), equalTo(defaultProtocols.size() + 1));
+        assertThat("JNLP3-connect must be enabled after the roundtrip", 
+                reloadedProtocols, hasItem("JNLP3-connect"));
+    }
+    
+    @Test
+    @Issue("JENKINS-39465")
+    public void agentProtocols_multipleDisable_roundtrip() throws Exception {
+        final Set<String> defaultProtocols = Collections.unmodifiableSet(j.jenkins.getAgentProtocols());
+        Assume.assumeThat("At least one protocol is enabled", defaultProtocols.size(), greaterThan(0));
+        
+        final String protocolToDisable = defaultProtocols.iterator().next();
+        
+        final Set<String> newProtocols = new HashSet<>(defaultProtocols);
+        newProtocols.remove(protocolToDisable);
+        j.jenkins.setAgentProtocols(newProtocols);
+        j.jenkins.save();
+        assertThat(protocolToDisable + " must be disabled before the roundtrip", 
+                j.jenkins.getAgentProtocols(), not(hasItem(protocolToDisable)));
+        final Set<String> agentProtocolsBeforeReload = j.jenkins.getAgentProtocols();
+        j.jenkins.reload();
+        
+        assertFalse("The protocol list must have been really refreshed", agentProtocolsBeforeReload == j.jenkins.getAgentProtocols());
+        assertThat("We should have disabled one protocol", 
+                j.jenkins.getAgentProtocols().size(), equalTo(defaultProtocols.size() - 1));
+        assertThat(protocolToDisable + " must be disabled after the roundtrip", 
+                j.jenkins.getAgentProtocols(), not(hasItem(protocolToDisable)));
+    }
+    
+    @Test
+    @Issue("JENKINS-39465")
+    public void agentProtocols_multipleEnable_roundtrip() throws Exception {
+        final Set<String> defaultProtocols = Collections.unmodifiableSet(j.jenkins.getAgentProtocols());
+        Assume.assumeThat("We assume that JNLP3-connect is disabled", 
+                defaultProtocols, not(hasItem("JNLP3-connect")));
+        Assume.assumeThat("We assume that JNLP4-connect is disabled", 
+                defaultProtocols, not(hasItem("JNLP4-connect")));
```

Comment:
This makes no sense: the test is skipped.

```xml
  <testcase name="agentProtocols_multipleEnable_roundtrip" classname="jenkins.model.JenkinsTest" time="0">
    <skipped message="We assume that JNLP4-connect is disabled: got: &lt;[CLI-connect, CLI2-connect, JNLP-connect, JNLP2-connect, JNLP4-connect, Ping]&gt;, expected: not a collection containing &quot;JNLP4-connect&quot;"/>
    <system-out><![CDATA[=== Starting agentProtocols_multipleEnable_roundtrip(jenkins.model.JenkinsTest)
]]></system-out>
    <system-err><![CDATA[Sep 27, 2017 2:31:48 PM org.jvnet.hudson.test.JenkinsRule createWebServer
INFO: Running on http://localhost:45889/jenkins/
Sep 27, 2017 2:31:48 PM jenkins.InitReactorRunner$1 onAttained
INFO: Started initialization
Sep 27, 2017 2:31:48 PM jenkins.InitReactorRunner$1 onAttained
INFO: Listed all plugins
Sep 27, 2017 2:31:48 PM jenkins.InitReactorRunner$1 onAttained
INFO: Prepared all plugins
Sep 27, 2017 2:31:48 PM jenkins.InitReactorRunner$1 onAttained
INFO: Started all plugins
Sep 27, 2017 2:31:48 PM jenkins.InitReactorRunner$1 onAttained
INFO: Augmented all extensions
Sep 27, 2017 2:31:49 PM jenkins.InitReactorRunner$1 onAttained
INFO: Loaded all jobs
Sep 27, 2017 2:31:49 PM jenkins.slaves.DeprecatedAgentProtocolMonitor initializerCheck
WARNING: This Jenkins instance uses deprecated Remoting protocols: CLI-connect,CLI2-connect,JNLP-connect,JNLP2-connectIt may impact stability of the instance. If newer protocol versions are supported by all system components (agents, CLI and other clients), it is highly recommended to disable the deprecated protocols.
Sep 27, 2017 2:31:49 PM jenkins.InitReactorRunner$1 onAttained
INFO: Completed initialization
Sep 27, 2017 2:31:49 PM jenkins.model.Jenkins cleanUp
INFO: Stopping Jenkins
Sep 27, 2017 2:31:49 PM jenkins.model.Jenkins cleanUp
INFO: Jenkins stopped
]]></system-err>
  </testcase>
```

Why would these tests need to `Assume` anything? There should be nothing environmental that would affect whether or not the test could be run: we are controlling this aspect of the test environment.

Reply:
agreed, the test needs to be reworked.

Reply:
Created https://issues.jenkins-ci.org/browse/JENKINS-47181